### ISSUE TITLE
pdfdict's `__getitem__` may raise an error when requesting indirect objects

### ISSUE THREAD

**Issue by sjamaan** [pmaupin/pdfrw#98]
I have some code to extract images from a PDF, which works as follows (excuse me for the long code):

```python
# Many thanks to
# http://stackoverflow.com/questions/2693820/extract-images-from-pdf-without-resampling-in-python
# (converted from PyPDF2 to pdfrw)
from PIL import Image
from io import BytesIO

def write_xobject_to_image(xobject, out):
	if xobject['/Subtype'] == '/Image':
		size = (xobject['/Width'], xobject['/Height']) # Sometimes(!) raises an error
		data = xobject.stream
		if isinstance(data, str):
			data = data.encode('latin1') # encoding??

		if xobject['/ColorSpace'] == '/DeviceRGB':
			mode = "RGB"
		else:
			mode = "P"

		filetype = None
		# This is stupid.  Sometimes it's a string, sometimes an array...
		if isinstance(xobject['/Filter'], str):
			filters = [xobject['/Filter']]
		else:
			filters = xobject['/Filter']

		if '/FlateDecode' in filters:
			filetype = 'png'
			img = Image.frombytes(mode, size, data)
			img.save(out)
		elif '/DCTDecode' in filters:
			filetype = 'jpg'
			out.write(data)
		elif '/JPXDecode' in filters:
			filetype = 'jp2'
			out.write(data)
		# else: error?

		# Stop for first image on page
		return filetype

	return None

# usage
from pdfrw import PdfReader
from pdfrw.objects import PdfName
from pdfrw.findobjs import find_objects

pdf_document = PdfReader(pdf_file)
for pageno in range(0, len(pdf_document.pages)):
	page = pdf_document.pages[pageno]

	images = find_objects(page, valid_subtypes=(PdfName.Image,))
	for img in images:
		with NamedTemporaryFile() as temp:
			write_xobject_to_image(img, temp)
```

Some PDFs will result in an error on the indicated line: `pdfrw.errors.PdfParseError: Dict key '/Height' is not a PdfName`

After some digging, it turns out that this is because `__getitem__()` in the `PdfDict` class accepts a string, which it passes to `get()`, which will convert it to  `PdfName`. Then if it turns out to be an indirect object, it will call `__setattr__` with the _original_ key, which is still a string, resulting in the aforementioned error.

Interestingly, if you change the code to add debugging info (like `print(xobject.keys())`), it will magically work (probably because indirect objects have already been resolved). I'm not sure how to fix this, nor how to even approach writing a test for this (I don't have much of a clue about how PDFs work), but this seems inconsistent: I'd expect the code to work, independent of whether the PDF has indirect objects or not.

I've now fixed this in my own code by removing all the square brackets, so it relies on `__getattr__` (which will always pass a `PdfName` to `get()`) instead of `__getitem__`. It now reads like this:

```python
def write_xobject_to_image(xobject, out):
	if xobject.Subtype == '/Image':
		size = (xobject.Width, xobject.Height)
		data = xobject.stream
		if isinstance(data, str):
			data = data.encode('latin1') # encoding??

		if xobject.ColorSpace == '/DeviceRGB':
			mode = "RGB"
		else:
			mode = "P"

		filetype = None
		# This is stupid.  Sometimes it's a string, sometimes an array...
		if isinstance(xobject.Filter, str):
			filters = [xobject.Filter]
		else:
			filters = xobject.Filter

		if '/FlateDecode' in filters:
			filetype = 'png'
			img = Image.frombytes(mode, size, data)
			img.save(out)
		elif '/DCTDecode' in filters:
			filetype = 'jpg'
			out.write(data)
		elif '/JPXDecode' in filters:
			filetype = 'jp2'
			out.write(data)
		# else: error?

		# Stop for first image on page
		return filetype

	return None
```

**Comment by pmaupin** [comment]
0) Thanks for the bug report!

1) I am sorry that string indexing seemed to sometime work -- that sort of bug is frustrating, indeed.  The indirect support was added several years after the start of the pdfrw project, and that problem was just missed.

2) But nobody else has reported this, probably because that is *not canonical pdfrw code*.  When using pdfrw, writing `if xobject['/Subtype'] == '/Image'` or even `if xobject.Subtype == '/Image'` is not canonical; instead write `if xobject.Subtype ==PdfName.Image`.

3) So I'm conflicted.  I could make your original code work by insuring that when an indirect object gets loaded, the key is coerced to be a `PdfName`, or I could just check and disallow non-PdfNames on `__getitem__` the way it is done on `__setitem__`  I suppose the first has the benefit of easing ports from PyPDF2 (but you'd still have the problem with `__setitem__`), while the second has the benefit of teaching the canonical way of doing things.

**Comment by sjamaan** [comment]
Hi,

Thanks for the quick and thoughtful reply!

> writing if xobject['/Subtype'] == '/Image' or even if xobject.Subtype == '/Image' is not canonical; instead write if xobject.Subtype ==PdfName.Image.

Good to know, I wasn't aware that that's the way to do this (and like I said, it's just a straight port of PyPDF2 code I didn't even write myself. I don't remember why I didn't want to go with PyPDF2).  However, the code didn't crash on that particular part of the code: it crashed on `size = (xobject['/Width'], xobject['/Height'])`. What would be the canonical way of expressing that?

**Comment by pmaupin** [comment]
I would just write:

`size = xobject.Width, xobject.Height`

Note that, for tuples in an assignment like this, you can use parentheses (and they are very useful if your tuple is so long it needs to span multiple lines), but they are not required.

**Comment by sjamaan** [comment]
That's what I ended up doing, but I still think the difference between bracket getters and attribute getters is unnecessarily confusing, especially due to the fact that if you do something that _should_ have no effect, like reading the keys, it suddenly changes the behaviour of the getter. Especially since that is typically one of the first things one would add when debugging, and when removing that you don't want the code to break.

Given that you're on the fence on what to do, I think the best approach would be to coerce the name to a `PdfName` object. The alternatives are either backwards-incompatible (if you decide to give an error when passed a string) or keeping the confusing status quo. If you don't mind breaking backwards compatibility, maybe it's best to drop the bracket notation altogether. Maybe I'm missing something, but to me it looks like redundant syntax for the same thing which nevertheless acts differently.

Thanks for pointing out that the tuple syntax doesn't require parentheses. Still, I find it clearer to add them anyway since there are situations where they _are_ required (in method calls, for example), and having a consistent syntax for them is less mental parsing overhead.

**Comment by pmaupin** [comment]
> maybe it's best to drop the bracket notation altogether

That, I'm not going to do, because, well, it *is* semantically a dict.  And, `foo[some_predefined_name]` is IMO clearer than `getattr(foo, some_predefined_name)`.

> The alternatives are either backwards-incompatible...

That's the thing.  Yes, it is backward compatible, but only for the getter.  The setter already has this check...  OTOH, people do a lot more getting than setting, so yeah, things will probably break.

> I think the best approach would be to coerce the name to a PdfName object.

I'm certainly leaning that way...

**Comment by Kanthavel-Securibox** [comment]
Hi,

I'm basically having the same issue here. The overwritten __get_item__ method seems to fail for some PDF documents.
I guess it relies on some lazy instantiation, since calling .keys() before __get_item__ solves the issue.
(I'm sorry I can't share any sample document but they all contain sensitive information).

Here's the code:

```
import io
from pdfrw import PdfReader

f = io.BytesIO(byte_array_content)  # The PDF byte content is stored in a bytearray 
reader = PdfReader(f)
```

(A) This fails (sometimes):
```
if y.pages and y.pages[0] and '/MediaBox' in y.pages[0] and len(y.pages[0].get('/MediaBox')) > 3:
#     do something
```

It raises the exception:
`pdfrw.errors.PdfParseError: Dict key '/MediaBox' is not a PdfName:`

(B) This works like a charm:
```
if y.pages and y.pages[0] and '/MediaBox' in y.pages[0].keys() and len(y.pages[0].get('/MediaBox')) > 3:
#     do something
```

Namely, after I have called (B), (A) starts working (I'm working in a jupyter notebook, jumping from one cell to the other).

Thanks a lot,

**Comment by pmaupin** [comment]
@Kanthavel-Securibox :

Semantically, dicts in PDF files are not the same as Python dicts, and pdfrw's PdfDict is designed to match a PDF dict's behavior as closely as possible.  One feature of PDF dicts is that there is no difference between a missing key and a key that has a null value.  pdfrw supports this by never raising a `KeyError` exception -- it returns `None` (pdfrw's version of a PDF null) instead.  So, instead of:

`if y.pages and y.pages[0] and '/MediaBox' in y.pages[0] and len(y.pages[0].get('/MediaBox')) > 3:`

You can write:

`if y.pages and y.pages[0] and len(y.pages[0].MediaBox or ()) > 3:`

**Comment by pmaupin** [comment]
Fixed in PR #108
